packages/playht/src/api/apiCommon.ts

```diff
@@ -284,5 +321,113 @@ async function audioStreamFromSentences(
     );
   });
 
-  promiseStream.pipe(writeAudio);
+  audioChunkStream.pipe(writeAudio);
+}
+
+class Task {
+  fn: Function
+  name: string
+
+  constructor(fn: Function, name: string) {
+    this.fn = fn;
+    this.name = name
+  }
 }
+
+/**
+ * Responsible for optimizing the rate at which text is sent to the underlying API endpoint, according to the
+ * specified {@link CongestionCtrl} algorithm.  {@link CongestionController} is essentially a task queue
+ * that throttles the parallelism of, and delay between, task execution.
+ *
+ * The primary motivation for this (as of 2024/02/28) is to protect PlayHT On-Prem appliances
+ * from being inundated with a burst of text-to-speech requests that it can't satisfy.  Prior to the introduction
+ * of {@link CongestionController} (and more generally {@link CongestionCtrl}), the client would split
+ * a text stream into two text chunks (referred to as "sentences") and send them to the API client (i.e. gRPC client)
+ * simultaneously.  This would routinely overload on-prem appliances that operate without a lot of GPU capacity headroom[1].
+ *
+ * The result would be that most requests that clients sent would immediately result in a gRPC error 8: RESOURCE_EXHAUSTED;
+ * and therefore, a bad customer experience.  {@link CongestionController}, if configured with {@link CongestionCtrl#StaticMar2024},
+ * will now delay sending subsequent text chunks (i.e. sentences) to the gRPC client until audio for the preceding text
+ * chunk has started streaming.
+ *
+ * The current {@link CongestionCtrl} algorithm ({@link CongestionCtrl#StaticMar2024}) is very simple and leaves a lot to
+ * be desired.  We should iterate on these algorithms.  The {@link CongestionCtrl} enum was added so that algorithms
+ * can be added without requiring customers to change their code much.
+ *
+ * [1] Customers tend to be very cost sensitive regarding expensive GPU capacity, and therefore want to keep their appliances
+ * running near 100% utilization.
+ *
+ * --mtp@2024/02/28
+ *
+ * This class is largely inert if the specified {@link CongestionCtrl} is {@link CongestionCtrl#Off}.
+ */
+class CongestionController {
+
+  algo: CongestionCtrl;
+  taskQ: Task[] = [];
+  inflight: number = 0;
+  parallelism: number;
+  postChunkBackoff: number;
+
+  constructor(algo: CongestionCtrl) {
+    this.algo = algo;
+    switch (algo) {
+      case CongestionCtrl.Off:
+        this.parallelism = Infinity;
+        this.postChunkBackoff = 0;
+        break;
+      case CongestionCtrl.StaticMar2024:
+        this.parallelism = 1;
```

Comment:
Short: Can we do this later?  I don't want to cause an outage by changing default behavior.

Long:

So the idea is that customers configure this by setting CongestionCtrl with PlayHT.init.  I'm wary of exposing the lower level knobs like parallelism and backoff directly to customers because I'm not convinced that static numbers are the right way to go medium-term.  Medium-term, I suspect we'll want a congestion control algorithm that adapts to feedback the client gets from turbo-prop (e.g. RESOURCE_EXHAUSTED, or actual latencies vs expected latencies).  If we expose the low level knobs directly to customers, then we'll have to accommodate those public low level knobs in perpetuity (or at least for some long deprecation time period).

I agree that parallelism = 2 is probably a good default (which is why I'm asking on-prem customers to use it); but I'm wary of imposing this behavior change on all customers, because it's not obvious to me what the performance artifacts will be in those cases.  I'm hand holding every on-prem customers through use-cases; both from the SDK side and the appliance side (e.g. adjust the number of GPUs they have available).  I suspect if we adjusted the parallelism here - we'd end up with a deluge of support requests (warranted or not) that we wouldn't be able to handle.

The way to do this would be to add a third CongestionCtrl option - say CongestionCtrl.SmoothApr2024 - that set the parallelism = 2 and postChunkBackoff=XY.  Then we could test it slowly with select customers or a % of random streams and see the results (assuming we also build a way to gather SDK telemetry).